src/apps/relay/windows/Service.cpp

```diff
@@ -0,0 +1,255 @@
+ * \param dwArgc   - number of command line arguments
+ * \param lpszArgv - array of command line arguments
+ */
+void WINAPI ServiceMain(DWORD dwArgc, LPTSTR lpszArgv[]) {
+  USES_CONVERSION;
+  g_Log("Enter ServiceMain");
+
+  // Register the handler function for the service
+  g_Service.handle = RegisterServiceCtrlHandler(g_Service.name, ServiceControlHandler);
+  if (!g_Service.handle) {
+    ServiceReportEvent(_T("RegisterServiceCtrlHandler fail"), GetLastError());
+    return;
+  }
+
+  // These SERVICE_STATUS members remain as set here
+  g_Service.status.dwServiceType = SERVICE_WIN32_OWN_PROCESS;
+  g_Service.status.dwServiceSpecificExitCode = 0;
+
+  // Report initial status to the SCM
+  ServiceReportStatus(SERVICE_START_PENDING, NO_ERROR, 3000);
+
+  // Perform service-specific initialization
+  if (g_Service.start) {
+    char **argv = new char *[dwArgc];
+    if (argv) {
+      for (DWORD i = 0; i < dwArgc; i++) {
+        argv[i] = T2A(lpszArgv[i]);
+      }
+      DWORD r = g_Service.start(dwArgc, argv);
+      delete[] argv;
+      if (ERROR_SUCCESS != r) {
+        ServiceReportEvent(_T("Service start fail"), r);
+        ServiceReportStatus(SERVICE_STOPPED, r, 0);
+        return;
+      }
+    }
+  }
+
+  // Report running status when initialization is complete.
+  ServiceReportStatus(SERVICE_RUNNING, NO_ERROR, 0);
+
+  // Perform service-specific work.
+  if (g_Service.run) {
+    DWORD r = g_Service.run();
+    if (ERROR_SUCCESS != r) {
+      ServiceReportEvent(_T("Service run fail"), r);
+      ServiceReportStatus(SERVICE_STOPPED, r, 0);
+      return;
+    }
+  }
+
+  ServiceReportStatus(SERVICE_STOPPED, NO_ERROR, 0);
+}
+
+/*!
+ * Run service
+ * \param name: the name of service
+ * \param start: the callback function of start. it maybe NULL
+ * \param run: the callback function of run. it maybe NULL
+ * \param stop: the callback function of stop. it maybe NULL
+ */
+int ServiceRun(char *name, fnServiceStart start, fnServiceRun run, fnServiceStop stop) {
+  USES_CONVERSION;
+  int nLen = 0;
+
+  g_Log("Enter ServiceRun");
+
+  if (!name) {
+    g_Log("Error: The name is NULL in ServiceRun");
+    return -1;
+  }
+
+  memset(&g_Service, 0, sizeof(WindowsService));
+  g_Service.start = start;
+  g_Service.run = run;
+  g_Service.stop = stop;
+  if (name) {
+    size_t nLen = strlen(name);
+    g_Service.name = new TCHAR[nLen + 1];
+    if (g_Service.name)
+      _tcsncpy_s(g_Service.name, nLen + 1, A2T(name), nLen);
+    else {
+      return -2;
+    }
+  }
+
+  // You can add any additional services for the process to this table.
+  SERVICE_TABLE_ENTRY dispatchTable[] = {{g_Service.name, (LPSERVICE_MAIN_FUNCTION)ServiceMain}, {NULL, NULL}};
+
+  // This call returns when the service has stopped.
+  // The process should simply terminate when the call returns.
+  if (!StartServiceCtrlDispatcher(dispatchTable)) {
+    ServiceReportEvent(_T("StartServiceCtrlDispatcher fail"), GetLastError());
+    nLen = GetLastError();
```

Comment:
if nLen represents a generic return value, it should have a better name.

Otherwise if it's supposed to be a length, it shouldn't be given the return value from GetLastError().